File: Part1_강의자료/Ch07_이분탐색/문제별코드/1654_랜선자르기/1654.cpp
#include <algorithm>
#include <stdio.h>
#include <vector>
using namespace std;

int calc_count(vector<long long> &lans, int len)
{
    long long cnt = 0;

    for (auto & it : lans)
        cnt += it / len;
    return cnt;
}

int main()
{
    int K, N;
    scanf("%d %d", &K, &N);

    vector<long long> lans(N);
    long long l = 1, r = -1;
    for (int i = 0; i < K; i++)
    {
        scanf("%lld", &lans[i]);
        r = max(r, lans[i]);
    }

    long long res = -1;
    while (l <= r)
    {
        long long mid = (l + r) / 2;
        if (calc_count(lans, mid) >= N)
        {
            l = mid + 1;
            res = mid;
        }
        else
            r = mid - 1;
    }
    printf("%lld\n", res);
    return 0;
}

File: Part1_강의자료/Ch07_이분탐색/문제별코드/1654_랜선자르기/1654.py
K, N = map(int, input().split())
lans = [int(input()) for _ in range(K)]

def calc_count(length):
    cnt = 0
    for lan in lans:
        cnt += (lan // length)
    return cnt

l, r = 1, max(lans)
res = -1
while l <= r:
    mid = (l + r) // 2
    if calc_count(mid) >= N:
        l = mid + 1
        res = mid
    else:
        r = mid - 1
print(res)
